cpp/src/parquet/column_reader.cc

```diff
@@ -857,7 +857,10 @@ class ColumnReaderImplBase {
   // first page with this encoding.
   void InitializeDataDecoder(const DataPage& page, int64_t levels_byte_size) {
     const uint8_t* buffer = page.data() + levels_byte_size;
-    const int64_t data_size = page.size() - levels_byte_size;
+    // PageReader may reuse the underlying buffer, so data_size
+    // should use `page.uncompressed_size() - levels_byte_size` rather
+    // than `page.size() - level_byte_size`.
+    const int64_t data_size = page.uncompressed_size() - levels_byte_size;
```

Comment:
nit: it would be helpful to comment this is now a precise uncompressed data size, compared to an upper bound in the past.

Reply:
Did you mean here: https://github.com/apache/arrow/blob/main/cpp/src/parquet/column_page.h#L53-L54 ?

It currently says `the total size in bytes of the page's data buffer` and it is kind of misleading. I agree we need to improve it. Something like `its size may exceed the actual size in use`? Please suggest a better statement @wjones127 

Or should we fix the `size()` function to always return the actual size in use to get rid of the confusion?

Reply:
So, here we should make `uncompressed_size` and `size` be equal (at least in DATA_PAGE V1?) I can fix it later this week

Reply:
Yes, that's where I meant. If there's some way to fix the `size()` function I would prefer that.

Reply:
Yeah I'm thinking could we instead fix change these lines:

https://github.com/apache/arrow/blob/99ac74e6890767e1a3cf96d7aff0b96b9503153c/cpp/src/parquet/column_reader.cc#L584-L588

To something like (change the inequality to `!=`):

```cpp
 // Grow the uncompressed buffer if we need to. Regardless, make sure the buffers
 // size is reported correctly.
 if (uncompressed_len != static_cast<int>(decompression_buffer_->size())) { 
   PARQUET_THROW_NOT_OK( 
       decompression_buffer_->Resize(uncompressed_len, /*shrink_to_fit=*/false)); 
 } 
```

Would that fix the same bug?